{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import os\n",
    "import random\n",
    "import subprocess\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH = \"~/data/paired/10x\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcr = ['IGH', 'IGL', 'IGK']\n",
    "tcr = ['TRA', 'TRB', 'TRG']\n",
    "trash = ['Multi', 'None']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# help functions\n",
    "\n",
    "def dict_to_sorted_df(dict_name, column_labels):\n",
    "    df = pd.DataFrame.from_dict(dict_name, orient='index')\n",
    "    df.columns = column_labels\n",
    "    return df.sort_index(0)\n",
    "\n",
    "def add_perc_to_df(dict_name, column_name):\n",
    "    dict_name['% ' + column_name] = dict_name['# '+column_name] / sum(dict_name['# '+column_name]) * 100\n",
    "    return dict_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_info_by_chain_type(filtered_contigs_info, chain_type):\n",
    "    return filtered_contigs_info.loc[filtered_contigs_info['chain'].isin(chain_type)]\n",
    "    \n",
    "\n",
    "def chains_simple_stats(filtered_contigs_info):\n",
    "    \n",
    "    # dict: chain_type -> number of contigs\n",
    "    contigs_types = {chain_type: len(rows) for chain_type, rows in filtered_contigs_info.groupby('chain')}\n",
    "    \n",
    "    stats_by_contig_types = dict_to_sorted_df(contigs_types, ['# contigs'])\n",
    "    stats_by_contig_types = add_perc_to_df(stats_by_contig_types, 'contigs')\n",
    "    \n",
    "    def summary_from_stats_by_contig_types(chain_type):\n",
    "        return sum(stats_by_contig_types.loc[stats_by_contig_types.index.isin(chain_type)]['% contigs'])\n",
    "    \n",
    "    summary = dict_to_sorted_df({\"bcr\" : [summary_from_stats_by_contig_types(bcr), \"ok\"], \n",
    "           \"tcr\" : [summary_from_stats_by_contig_types(tcr), \"filter out\"], \n",
    "           \"trash\" : [summary_from_stats_by_contig_types(trash), \"igblast for fun + filter out\"]}, [\"% contigs\", \"TODO\"])\n",
    "    \n",
    "    # print stats\n",
    "    print(\"> Number of all contigs: {}\\n\".format(sum(stats_by_contig_types['# contigs'])))\n",
    "    \n",
    "    print(\"> Contigs by chain types:\\n\")\n",
    "    print(stats_by_contig_types)\n",
    "    print(\"\\n(A value of Multi indicates that segments from multiple chains were present)\")\n",
    "    \n",
    "    print(\"\\n> Summary:\\n\")\n",
    "    print(summary)\n",
    "    \n",
    "\n",
    "def cells_simple_stats(bcr_contigs_info):\n",
    "    \n",
    "    # dict : cell_barcode -> number of bcr chains\n",
    "    cells_types = {cell_barcode : len(rows) for cell_barcode, rows in bcr_contigs_info.groupby('barcode')}\n",
    "    \n",
    "    summary = dict_to_sorted_df(dict(Counter(cells_types.values())), ['# cells'])\n",
    "    summary = add_perc_to_df(summary, 'cells')\n",
    "    summary.index = [str(name) + \" chains\" for name in summary.index]\n",
    "     \n",
    "    # dict: cell_barcode -> counts of bcr chains \n",
    "    chains_counts_by_cell = {cell_barcode : str(dict(Counter(rows['chain']))) for cell_barcode, rows in bcr_contigs_info.groupby('barcode')}\n",
    "    accurate_counts = dict_to_sorted_df(dict(Counter(chains_counts_by_cell.values())), ['# cells'])\n",
    "    accurate_counts = accurate_counts.sort_values(by='# cells',  ascending=False)\n",
    "    accurate_counts = add_perc_to_df(accurate_counts, 'cells')\n",
    "    \n",
    "    \n",
    "    # print stats\n",
    "    print(\"> Number of all cells: {}\\n\".format(len(cells_types)))\n",
    "    \n",
    "    print(\"> Cells by number of all chains\\n\")\n",
    "    print(summary)\n",
    "    \n",
    "    print(\"\\n> Cells by combination of chains\\n\")\n",
    "    print(accurate_counts)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_filtered_contigs_csv(dataset_name):\n",
    "    \n",
    "    annotation_fname = PATH + '/{}/vdj_v1_{}_b_filtered_contig_annotations.csv'.format(dataset_name, dataset_name)\n",
    "    \n",
    "    # read filtered_contig_annotations.csv\n",
    "    filtered_contigs_info = pd.read_csv(annotation_fname, delimiter=',')\n",
    "    \n",
    "    print(\"\\n\\n=========== contigs ===========\\n\\n\")\n",
    "    chains_simple_stats(filtered_contigs_info)\n",
    "    \n",
    "    print(\"\\n\\n=========== bcr chains ===========\\n\\n\")    \n",
    "    bcr_contigs_info = get_info_by_chain_type(filtered_contigs_info, bcr)    \n",
    "    cells_simple_stats(bcr_contigs_info)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CD19"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CD19+ B cells isolated from peripheral blood mononuclear cells (PBMCs) of a healthy donor, purchased from AllCells (Catalog #: PB010-0), 93% viable by Trypan blue stain.\n",
    "\n",
    "CD19+ B cells are primary cells with relatively small amounts of RNA (~1pg RNA/cell).\n",
    "\n",
    "Libraries were prepared following the Single Cell V(D)J Reagent Kits User Guide (CG000086 RevC).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "=========== contigs ===========\n",
      "\n",
      "\n",
      "> Number of all contigs: 33486\n",
      "\n",
      "> Contigs by chain types:\n",
      "\n",
      "       # contigs  % contigs\n",
      "IGH        11074  33.070537\n",
      "IGK         7234  21.603058\n",
      "IGL         5191  15.502001\n",
      "Multi       7488  22.361584\n",
      "None           4   0.011945\n",
      "TRA          190   0.567401\n",
      "TRB         2296   6.856597\n",
      "TRG            9   0.026877\n",
      "\n",
      "(A value of Multi indicates that segments from multiple chains were present)\n",
      "\n",
      "> Summary:\n",
      "\n",
      "       % contigs                          TODO\n",
      "bcr    70.175596                            ok\n",
      "tcr     7.450875                    filter out\n",
      "trash  22.373529  igblast for fun + filter out\n",
      "\n",
      "\n",
      "=========== bcr chains ===========\n",
      "\n",
      "\n",
      "> Number of all cells: 9467\n",
      "\n",
      "> Cells by number of all chains\n",
      "\n",
      "          # cells    % cells\n",
      "1 chains       81   0.855604\n",
      "2 chains     5815  61.423894\n",
      "3 chains     2737  28.910954\n",
      "4 chains      639   6.749762\n",
      "5 chains      156   1.647829\n",
      "6 chains       33   0.348579\n",
      "7 chains        5   0.052815\n",
      "8 chains        1   0.010563\n",
      "\n",
      "> Cells by combination of chains\n",
      "\n",
      "                                # cells    % cells\n",
      "{'IGK': 1, 'IGH': 1}               3672  38.787367\n",
      "{'IGL': 1, 'IGH': 1}               2121  22.404141\n",
      "{'IGL': 1, 'IGK': 1, 'IGH': 1}      792   8.365903\n",
      "{'IGK': 1, 'IGH': 2}                652   6.887081\n",
      "{'IGK': 2, 'IGH': 1}                501   5.292067\n",
      "{'IGL': 2, 'IGH': 1}                476   5.027992\n",
      "{'IGL': 1, 'IGH': 2}                314   3.316785\n",
      "{'IGL': 1, 'IGK': 1, 'IGH': 2}      229   2.418929\n",
      "{'IGK': 2, 'IGH': 2}                139   1.468258\n",
      "{'IGL': 2, 'IGK': 1, 'IGH': 1}       97   1.024612\n",
      "{'IGL': 2, 'IGH': 2}                 82   0.866167\n",
      "{'IGL': 1, 'IGK': 2, 'IGH': 1}       61   0.644344\n",
      "{'IGL': 1, 'IGK': 2, 'IGH': 2}       47   0.496461\n",
      "{'IGK': 1}                           46   0.485898\n",
      "{'IGL': 2, 'IGK': 1, 'IGH': 2}       42   0.443646\n",
      "{'IGL': 1, 'IGK': 1, 'IGH': 3}       29   0.306327\n",
      "{'IGL': 1}                           25   0.264075\n",
      "{'IGK': 1, 'IGH': 3}                 12   0.126756\n",
      "{'IGH': 1}                           10   0.105630\n",
      "{'IGK': 2}                           10   0.105630\n",
      "{'IGL': 3, 'IGH': 1}                  9   0.095067\n",
      "{'IGK': 2, 'IGH': 3}                  9   0.095067\n",
      "{'IGL': 2, 'IGK': 2, 'IGH': 1}        7   0.073941\n",
      "{'IGL': 1, 'IGK': 1}                  7   0.073941\n",
      "{'IGK': 3, 'IGH': 2}                  7   0.073941\n",
      "{'IGL': 1, 'IGH': 3}                  6   0.063378\n",
      "{'IGL': 1, 'IGK': 2, 'IGH': 3}        6   0.063378\n",
      "{'IGL': 3, 'IGK': 1, 'IGH': 2}        6   0.063378\n",
      "{'IGL': 2, 'IGH': 3}                  6   0.063378\n",
      "{'IGL': 2, 'IGK': 1, 'IGH': 3}        5   0.052815\n",
      "{'IGL': 2, 'IGK': 2, 'IGH': 2}        5   0.052815\n",
      "{'IGL': 2}                            4   0.042252\n",
      "{'IGL': 1, 'IGK': 3, 'IGH': 2}        4   0.042252\n",
      "{'IGL': 3, 'IGH': 2}                  3   0.031689\n",
      "{'IGL': 3, 'IGK': 1, 'IGH': 3}        3   0.031689\n",
      "{'IGL': 4, 'IGH': 2}                  3   0.031689\n",
      "{'IGL': 1, 'IGK': 3, 'IGH': 1}        3   0.031689\n",
      "{'IGK': 3, 'IGH': 1}                  3   0.031689\n",
      "{'IGL': 2, 'IGK': 1}                  2   0.021126\n",
      "{'IGL': 3, 'IGH': 3}                  2   0.021126\n",
      "{'IGL': 3, 'IGK': 1, 'IGH': 1}        2   0.021126\n",
      "{'IGK': 4, 'IGH': 2}                  2   0.021126\n",
      "{'IGL': 2, 'IGK': 2, 'IGH': 3}        1   0.010563\n",
      "{'IGL': 2, 'IGK': 2}                  1   0.010563\n",
      "{'IGL': 2, 'IGK': 3, 'IGH': 3}        1   0.010563\n",
      "{'IGH': 2}                            1   0.010563\n",
      "{'IGL': 1, 'IGH': 4}                  1   0.010563\n",
      "{'IGL': 1, 'IGK': 2, 'IGH': 4}        1   0.010563\n"
     ]
    }
   ],
   "source": [
    "process_filtered_contigs_csv('cd19')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GM12878"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "B-lymphoblastoid cell line GM12878, purchased from Coriell (Catalog #: GM12878), 87% viable by Trypan blue stain.\n",
    "\n",
    "GM12878 cells are a B-lymphoblastoid cell line with high expression level of Ig transcripts (mostly IGHM/D and IGL isotype).\n",
    "\n",
    "Libraries were prepared following the Single Cell V(D)J Reagent Kits User Guide (CG000086 RevC)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "=========== contigs ===========\n",
      "\n",
      "\n",
      "> Number of all contigs: 2711\n",
      "\n",
      "> Contigs by chain types:\n",
      "\n",
      "       # contigs  % contigs\n",
      "IGH         1019  37.587606\n",
      "IGK           17   0.627075\n",
      "IGL         1188  43.821468\n",
      "Multi        447  16.488381\n",
      "TRA            9   0.331981\n",
      "TRB           31   1.143489\n",
      "\n",
      "(A value of Multi indicates that segments from multiple chains were present)\n",
      "\n",
      "> Summary:\n",
      "\n",
      "       % contigs                          TODO\n",
      "bcr    82.036149                            ok\n",
      "tcr     1.475470                    filter out\n",
      "trash  16.488381  igblast for fun + filter out\n",
      "\n",
      "\n",
      "=========== bcr chains ===========\n",
      "\n",
      "\n",
      "> Number of all cells: 865\n",
      "\n",
      "> Cells by number of all chains\n",
      "\n",
      "          # cells    % cells\n",
      "1 chains       14   1.618497\n",
      "2 chains      475  54.913295\n",
      "3 chains      259  29.942197\n",
      "4 chains      102  11.791908\n",
      "5 chains       15   1.734104\n",
      "\n",
      "> Cells by combination of chains\n",
      "\n",
      "                                # cells    % cells\n",
      "{'IGL': 1, 'IGH': 1}                463  53.526012\n",
      "{'IGL': 2, 'IGH': 1}                211  24.393064\n",
      "{'IGL': 2, 'IGH': 2}                 99  11.445087\n",
      "{'IGL': 1, 'IGH': 2}                 42   4.855491\n",
      "{'IGL': 2, 'IGH': 3}                 10   1.156069\n",
      "{'IGL': 1}                            8   0.924855\n",
      "{'IGL': 2}                            7   0.809249\n",
      "{'IGH': 1}                            6   0.693642\n",
      "{'IGL': 1, 'IGK': 1, 'IGH': 1}        6   0.693642\n",
      "{'IGL': 2, 'IGK': 1, 'IGH': 2}        5   0.578035\n",
      "{'IGK': 1, 'IGH': 1}                  4   0.462428\n",
      "{'IGL': 2, 'IGK': 1, 'IGH': 1}        2   0.231214\n",
      "{'IGH': 2}                            1   0.115607\n",
      "{'IGL': 1, 'IGH': 3}                  1   0.115607\n"
     ]
    }
   ],
   "source": [
    "process_filtered_contigs_csv('gm12878')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
